SurvivalProbability - Readability, performance, and algorithm changes

bieniekmateusz opened the pull request:
Small changes:
- Performance (16 minutes to 60 seconds): 
-- issue due to the unnecessary function overhead,  
-- iterating over the dataset only once (RAM memory optimisation)
- More thorough documentation of parameters and additional sanity checks in __init__
- Removed temporarily the progress meter which only worked for loading the simulation dataset, confusing the reader
- Allows the user to access the distribution for each tau through sp.sp_timeseries
- Returns the taus along with the timeseries (opening access for dt). Not returning tau = 0 

Validity Test Cases Changes:
- Replaced the arbitrary test cases with defined datasets and predictable taus
- Added a test case where no atoms IDs are found. In this case, we return tau = NaN. This is because 0 means that the initial molecular leaves after a given tau. However, if there is no initial molecules in the first place, this would be the wrong conclusion. 

Algorithmic Changes made in this Pull Request - Request for Contribution from the Original Author: 
 - We modified the algorithm to be applicable to other atom groups (e.g. ions) besides water. Particularly, we changed the way the algorithm behaves in the case where there is no molecules found for the reference frame. Previously, one would expect the reference frame to always find some water molecules. If this was not the case, each tau was diluted - divided by (Nt + 1). Because it was unlikely to happen with any reasonable water selection, this case was not important. However, let us consider using SP the survival of ions in a given area, with frames where no ions are found in the first place (reference frame). For such frames, diluting the taus means that the ions leave earlier - whereas in fact, these ions might rarely diffuse into the selected area. The rare diffusion into the selected area should not affect how long they stay (ie their SurvivalProbability). Please let us know if we missed anything. 

Questions:
 - Since this algorithm is applicable to any mobile atoms in a system, we suggest this package is moved out of waterdynamics into a more general package. 

Future considerations:
- Adding progress bar
- Adding dt to allow sample the dataset with a faster shift (and possibly discuss data correlations for larger taus)

Edit: Acknowledgement - This work was done together with @p-j-smith, but we submitted most of it on my account. 

PR Checklist
------------
 - [X ] Tests?
 - [X] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

richardjgowers commented:
@bieniekmateusz thanks, this looks like it will make some good changes.

WRT "faster shift for dt", are you talking about doing every 2nd frame rather than every frame?

richardjgowers commented:
For the case where no molecules are found, it makes sense to disregard this run as you can't measure anything.  It's not that the molecule had 0 survival probability.

So for [hydrogen bond autocorrelation](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/hbonds/hbond_autocorrel.py#L208) I let people define a total duration (tau_max here) and a number of frames within that duration to consider.  I'm not completely happy with that solution either, so it would be good to define a clear & comprehensive way/language to define this sort of stuff.

I need to read through this module fully but..

Reducing the resolution of frames should be fine in some cases.  We can actually check after the calculation whether the calculated tau makes sense with respect to how we sampled it.  Then issue some sort of warning if the data is probably junk (and suggest better settings).

Overlapping frames should be fine.  The concentration in an area will remain constant, so by starting a new sample in the same window will identify different "starting molecules".  It will be slightly correlated with the old sample though (so isn't 100% new data)

bieniekmateusz commented:
I've updated the last test case (t0) to avoid arbitrary datasets and to test tf in the same test case.

bieniekmateusz commented:
Do I understand it right that the error messages are not related to our submission? Please let us know if you we should take any actions regarding this. Thanks

bieniekmateusz commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1197,96 +1199,73 @@ class SurvivalProbability(object):
+        self.sp_timeseries = sp_timeseries
+
+        return tau_timeseries, [np.mean(sp) for sp in sp_timeseries]
 
```
Only now that you pointed it out I've noticed. Corrected and pushed. Thanks

bieniekmateusz commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -412,6 +412,8 @@
 """
 from __future__ import print_function, division, absolute_import
+
+from numpy.core.multiarray import ndarray
```
Deleted, it was not used anymore.

bieniekmateusz commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1197,96 +1199,72 @@ class SurvivalProbability(object):
-            output = self._getMeanOnePoint(selected, window_size)
-            self.timeseries.append(output)
-
+        self.tf = len(list(universe.trajectory[:tf])) + 1
```
Just tested it and the `list` is prohibitively expensive as pointed out. Will correct while switching to the run(start, stop, step)

bieniekmateusz commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1197,96 +1199,72 @@ class SurvivalProbability(object):
+            selected.append(set(self.universe.select_atoms(self.selection).ids))
 
-        return sumDeltaP/n
+        tau_timeseries = list(range(1, self.tau_max + 1))
```
Done.

bieniekmateusz commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1197,96 +1199,72 @@ class SurvivalProbability(object):
 
-        return sumDeltaP/n
+        tau_timeseries = list(range(1, self.tau_max + 1))
+        sp_timeseries = [[] for _ in range(self.tau_max)]
```
So we do not know how many datapoints we are going to have ahead. This is because some reference frames will not contain the original molecule - hence they would be omitted.

bieniekmateusz commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1197,96 +1199,72 @@ class SurvivalProbability(object):
+            print(args)
 
-    def _getMeanOnePoint(self, selected, window_size):
+    def run(self):
```
Thanks, Would you also recommend us to do the classical parameter documentation in the run function? ie
        :param start: means ..
        :param stop: documentation
...

This would be read by any semi-smart code editor I think.

richardjgowers commented:
Yeah but we use numpy style docs -
https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/rdf.py
On Mon, Jul 23, 2018 at 7:09 PM, Mateusz Bieniek <notifications@github.com>
wrote:

> *@bieniekmateusz* commented on this pull request.
> ------------------------------
>
> In package/MDAnalysis/analysis/waterdynamics.py
> <https://github.com/MDAnalysis/mdanalysis/pull/1995#discussion_r204590134>
> :
>
> >
> -    def _getMeanOnePoint(self, selected, window_size):
> +    def run(self):
>
> Thanks, Would you also recommend us to do the classical parameter
> documentation in the run function? ie
> :param start: means ..
> :param stop: documentation
> ...
>
> This would be read by any semi-smart code editor I think.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1995#discussion_r204590134>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jB4ZKnsBZLLij2FZl9qrmMKMgTt4Fks5uJmWngaJpZM4VUo10>
> .
>

bieniekmateusz commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1197,96 +1199,72 @@ class SurvivalProbability(object):
+        # load all frames to an array of sets
+        selected = []
+        for _ in self.universe.trajectory[self.t0:self.tf]:
+            selected.append(set(self.universe.select_atoms(self.selection).ids))
```
We've added the `.ids` to ensure that we keep the minimum amount of data for the analysis - although I am not sure if we helped much this way. The `set` is required in order to use the set.intersection later. 

I've run a few quick performance tests but 99.9% of time is spent loading the frames.

kain88-de commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1197,96 +1199,72 @@ class SurvivalProbability(object):
-        # fraction of water molecules that survived
-        Ntau = self._NumPart_tau(selected, t, tau)
-        return Ntau/Nt
+                Ntau = len(set.intersection(*selected[t:t + tau + 1]))
```
https://www.mdanalysis.org/docs/documentation_pages/core/groups.html#MDAnalysis.core.groups.AtomGroup

This is documented in the common groups docs. You have to scroll down on that page

kain88-de commented on package/CHANGELOG:
```diff
@@ -53,6 +53,8 @@ Changes
     *Group.unique will always return the same object unless the group is
     updated or modified. (PR #1922)
+  * SurvivalProbability does not dilute SP anymore when the reference frame (t0)
+    cannot find any molecules in the first place. (PR #1995)
```
Can you also add both of your names to the author list contributing to this release at the top. Thanks.

bieniekmateusz replied:
I assme you mean just above the  0.18.1 in CHANGELOG. Done

kain88-de replied:
Yes just above the 0.18.1. I'm not sure if you are done yet. Github doesn't show the new names. We tend to use github names for that list.

bieniekmateusz commented:
I've updated the code and introduced the new interface. Also, I updated the documentation to mirror the changes in the code. 

Unfortunately, `step` does not help with the performance increase. I'll give it some thought and improve it in the future hopefully. I have an idea how to tackle that but it will be another PR.

bieniekmateusz commented:
I further updated the run interface to be consistent with the #1463. However, should have I used the **kwargs instead, let me know. Thanks

richardjgowers commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1194,99 +1192,88 @@ class SurvivalProbability(object):
+    def __init__(self, universe, selection, verbose=False):
         self.universe = universe
         self.selection = selection
-        self.t0 = t0
```
So we can't just stop allowing `t0` and `tf` here, because it would break people's existing code.  What we need to do instead is to allow it, but raise warnings if it's used, so something like:


```python
def __init__(...., t0=None, tf=None):
    if not t0 is None:
        self.start = t0
        warnings.warn("t0 is deprecated, use run(start=t0) instead", category=DeprecationWarning)

bieniekmateusz replied:
Updated, thanks.

bieniekmateusz commented:
Should I reverse the CHANGELOG file in order to have the run the tests again? 

Also, let me know if there is anything else I should address. 

Thanks, Mat

kain88-de commented:
It would be nice if you can rebase against develop. Then the tests will run
again.

On Fri, Jul 27, 2018 at 3:39 PM Mateusz Bieniek <notifications@github.com>
wrote:

> Should I reverse the CHANGELOG file in order to have the run the tests
> again?
>
> Also, let me know if there is anything else I should address.
>
> Thanks, Mat
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1995#issuecomment-408421165>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEGnVoD-SuKmK7tZrhM9zV52_hYBq3-wks5uKxf2gaJpZM4VUo10>
> .
>

bieniekmateusz commented:
Can I use a simple merge for this? Rebase seems to me like one evening of "pure joy".

richardjgowers commented:
@bieniekmateusz if you update your local develop branch (fetch + pull) then just do `git rebase develop` you should just have to fix CHANGELOG, which will be including both your addition + whatever else is new since you branched from develop.  Shouldn't be too painful

richardjgowers commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1194,99 +1194,108 @@ class SurvivalProbability(object):
+            raise ValueError("\"stop\" must be smaller than the number of frames in the trajectory.")
 
+        if stop == -1:
+            stop = self.universe.trajectory[-1].frame + 1
```
this will load the frame, aka slow.  Can you just use `None` for stop?  Or maybe `len(u.trajectory)` when you need to check bounds?

bieniekmateusz replied:
Sounds much better. Done

richardjgowers commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1194,99 +1194,108 @@ class SurvivalProbability(object):
-        return sumDeltaP/n
+        # load all frames to an array of sets
+        selected_ids = []
+        for ts in self.universe.trajectory[start:stop]:
```
add `step`?

richardjgowers replied:
ah, I see you're just using step below.  So every frame is a start point, but you only do deltas of `step`?

bieniekmateusz replied:
Hmm, not sure if I get the question. So this step here is tricky. I first load all frames (loading the right ones and considering the step makes it very tricky). 

And then during the analysis, I start the gathering the tau every step. So assume step=5 and max tau is 3. We'll have:

t = 0:
t + taus: 0, 1, 2, 3
t = 5:
t + taus: 5, 6, 7, 8, 
t= 10: .. etc

The main speed gain would be from the ensuring that frame 4, 9, ... etc would not be loaded. I will think about this how to get this right and not overcomplicate the algorithm.

richardjgowers commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1194,99 +1194,108 @@ class SurvivalProbability(object):
+        # user can investigate the distribution and sample size
+        self.sp_timeseries = sp_timeseries
+
+        return tau_timeseries, [np.mean(sp) for sp in sp_timeseries]
```
It's better here to save the attributes to the class and return the class from `run` (`return self`)

kain88-de replied:
*better* means that it follows the behavior of other analysis classes in MDAnalysis.  The general pattern to use *any* analysis class in MDAnalysis is this.

```python
sp = SurvivalProbability(....).run()
sp.tau
sp.means
```

If we keep all classes to follow this it will make MDAnalysis easier/more intuitive to use.

bieniekmateusz replied:
Thanks @richardjgowers and @kain88-de 

I've updated the code to conform to your comments. 

If you don't mind I will this opportunity to share some thoughts. 1) @kain88-de The constructor in python has to return None, 2) I have not seen a single API that returns data in the object's field, I imagine that you have some bigger picture in mind that I missing,  3) If the purpose of returning self is to be able to use again .run() straight away, that would overwrite the previous data. 

I'll update the comments/documentation once all the code is reviewed. Thanks

kain88-de replied:
1 ) I'm not sure what you mean. But yeah the `__ini__` function doesn't return anything. But `Object()` does return an instance of the class after `__init__` was called.
2 ) See `rmsd.RMSD`, `rmsd.RMSF`, `contacts.Contacts`. and other analysis classes using the base analysis class store results in `self`. 
3 )  It allows writing a [base class](https://github.com/MDAnalysis/mdanalysis/blob/461aadf6733d234e0913bae774a6312a208ede27/package/MDAnalysis/analysis/base.py#L44) that shares reusable code between different analysis types. Having `run` return itself allows method chaining and code I've written above. Mainly it fits with how the rest of analysis classes work. Consistency in API is key here.

bieniekmateusz replied:
Oh, of course you're right about 1). Sorry, my mistake. Thanks, I'll have a look at the base class too. Let me know if you see anything else in the code that needs attention. Mat

bieniekmateusz commented:
I did the rebase which was not too much trouble at the end. Cheers

bieniekmateusz commented:
Error in testsuite/MDAnalysisTests/topology/test_pqr.py:93: 

Could it be that the rebasing introduced this issue since I have not touched the topology files. Do you have any suggestions on what I should do to correct the tests? Thanks

bieniekmateusz commented:
I've updated the example to match the data being stored in the object's fields.

bieniekmateusz commented:
and @p-j-smith! 

We're happy to hear that. And thanks to you too!
